### PR TITLE
External replication refactoring

### PR DESCRIPTION
**Pull request by noname0443**
I want to make "strategy" pattern with `IExternalReplication` in `app`. To reach this purpose I made `UnimplementedExternalReplication` which contain empty methods and `ExternalReplication` which contain `Node` methods.

### CODE REVIEW

File: internal/app/app.go [hunk 1]
```diff
@@ -26,17 +27,18 @@ import (
...
+	config              *config.Config
+	dcs                 dcs.DCS
+	cluster             *mysql.Cluster
+	filelock            *flock.Flock
+	nodeFailedAt        map[string]time.Time
+	streamFromFailedAt  map[string]time.Time
+	daemonState         *DaemonState
+	daemonMutex         sync.Mutex
+	replRepairState     map[string]*ReplicationRepairState
+	ExternalReplication replication.IExternalReplication
```
Comment (Fizic) [inline internal/app/app.go:41]: Maybe it's better not to add this field to the app and add it to the node

**Review by Fizic — CHANGES_REQUESTED** [review]

File: internal/config/config.go [hunk 2]
```diff
@@ -87,6 +87,7 @@ type Config struct {
 	TestFilesystemReadonlyFile              string              `config:"test_filesystem_readonly_file" yaml:"test_filesystem_readonly_file"`
 	ReplicationChannel                      string              `config:"replication_channel" yaml:"replication_channel"`
 	ExternalReplicationChannel              string              `config:"external_replication_channel" yaml:"external_replication_channel"`
+	IsExternalReplicationSupported          bool                `config:"is_external_replication_supported" yaml:"is_external_replication_supported"`
```
[the remark this answers is not in the document]
Comment (noname0443) [reply internal/config/config.go:90]: `CheckIfExternalReplicationSupported` need a `version` object. To get version we have to make query. If we don't use external replication we anyway have to make query. I think we need something to hard turn off external replication by default if we don't really need in it. In this case we can save resources.

File: internal/mysql/queries.go [hunk 3]
```diff
@@ -1,76 +1,76 @@
...
-	queryGetExternalReplicationSettings = "get_external_replication_settings"
-	queryChangeSource                   = "change_source"
-	queryResetReplicaAll                = "reset_replica_all"
-	queryStopReplica                    = "stop_replica"
-	queryStartReplica                   = "start_replica"
-	queryIgnoreDB                       = "ignore_db"
-	querySetInnodbFlushLogAtTrxCommit   = "set_innodb_flush_log_at_trx_commit"
-	querySetSyncBinlog                  = "set_sync_binlog"
-	queryGetReplicationSettings         = "get_replication_settings"
+	QueryPing                           = "ping"
```
[the remark this answers is not in the document]
Comment (noname0443) [reply internal/mysql/queries.go:4]: I need access to queries in internal/app/replication/mysql.go, but I can evade problem if I duplicate `queryGetExternalReplicationSettings` for external replication implementation. How do you think it is a good idea?

File: internal/app/app.go (hunk 1 above)
Comment (noname0443) [reply internal/app/app.go:41]: I want to make "strategy" pattern for `IExternalReplication` in `app`. In this case I think we can't use `Node` to reach purpose.

File: internal/app/replication/replication.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+type IExternalReplication interface {
+	IsSupported(*mysql.Node) (bool, error)
+	Set(*mysql.Node) error
+	Reset(*mysql.Node) error
+	Start(*mysql.Node) error
+	Get(*mysql.Node) (mysql.ReplicaStatus, error)
+	Stop(*mysql.Node) error
+}
+
+type DummyExternalReplication struct{}
```
[the remark this answers is not in the document]
Comment (noname0443) [reply internal/app/replication/replication.go:14]: Let's rename it `DummyExternalReplication` -> `UnimplementedExternalReplication` and embed it to ExternalReplication.

File: internal/app/app.go [hunk 5]
```diff
@@ -229,7 +236,7 @@ func (app *App) externalCAFileChecker(ctx context.Context) {
 		select {
 		case <-ticker.C:
 			localNode := app.cluster.Local()
-			replicaStatus, err := localNode.GetExternalReplicaStatus()
+			replicaStatus, err := app.externalReplication.Get(localNode)
```
Comment (teem0n) [inline internal/app/app.go:239]: Get => GetReplicaStatus

File: internal/app/replication/replication.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Fizic) [reply internal/app/replication/replication.go]: It seems to me that while this is not being used, it should be removed. Also, a separate implementation is used for tests, usually via gomock

File: internal/app/replication/mysql.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	IsExternalReplicationSupported bool
+	logger                         *log.Logger
+}
+
+func NewExternalReplication(IsExternalReplicationSupported bool, logger *log.Logger) (*ExternalReplication, error) {
+	return &ExternalReplication{
+		IsExternalReplicationSupported: IsExternalReplicationSupported,
+		logger:                         logger,
+	}, nil
+}
```
Comment (teem0n) [inline internal/app/replication/mysql.go:21]: Here we should return ExternalReplication{} or UnimplementedExternalReplication{} depending on IsExternalReplicationSupported setting.

File: internal/app/replication/mysql.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+import (
+	"database/sql"
+
+	"github.com/yandex/mysync/internal/log"
+	"github.com/yandex/mysync/internal/mysql"
+)
+
+type ExternalReplication struct {
+	UnimplementedExternalReplication
+	IsExternalReplicationSupported bool
```
Comment (teem0n) [inline internal/app/replication/mysql.go:12]: This field should be removed. If we use ExternalReplication it means that setting is True

File: internal/app/replication/mysql.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+import (
+	"database/sql"
+
+	"github.com/yandex/mysync/internal/log"
+	"github.com/yandex/mysync/internal/mysql"
+)
+
+type ExternalReplication struct {
+	UnimplementedExternalReplication
```
Comment (teem0n) [inline internal/app/replication/mysql.go:11]: This should be removed

**Review by teem0n — CHANGES_REQUESTED** [review]

File: internal/app/app.go [hunk 9]
```diff
@@ -229,7 +235,7 @@ func (app *App) externalCAFileChecker(ctx context.Context) {
 		select {
 		case <-ticker.C:
 			localNode := app.cluster.Local()
-			replicaStatus, err := localNode.GetExternalReplicaStatus()
+			replicaStatus, err := app.externalReplication.GetReplicaStatus(localNode)
```
Comment (teem0n) [inline internal/app/app.go:238]: If we have External Replication turned off - we should not run externalCAFileChecker goroutine

**Review by teem0n — CHANGES_REQUESTED** [review]

**Review by teem0n — APPROVED** [review]

**Review by Fizic — APPROVED** [review]

**Pull request merged**